src/editor/codemirror/edits.ts

```diff
@@ -122,7 +124,65 @@ export const calculateChanges = (
   if (importInsertPoint.used) {
     changes[0].insert = importInsertPoint.prefix + changes[0].insert;
   }
+
+  if (userEvent === callableCode) {
+    //adjusts for closing bracket and newline char
+    const callableAdjustment = 2;
+
+    //point of code added, plus code length
+    additionInsertPoint =
+      changes[changes.length - 1].from +
+      changes[changes.length - 1].insert.length -
+      callableAdjustment;
+
+    //if import statement added
+    //assumes that there may be more than one import
+    if (changes.length > 1) {
+      additionInsertPoint += changes
+        .slice(0, changes.length - 1)
+        .flatMap((c) => c.insert)
+        .join("").length;
+    }
+  }
+
+  if (userEvent === exampleCode) {
+    if (changes.length > 1) {
+      //if import statement added
+      //assumes that there may be more than one import
+      const fromOffset = changes
+        .flatMap((c) => c.from)
+        .reduce((acc, cur) => acc + cur, 0);
+
+      //from offset is zero if code is dragged into empty editor
+      if (!fromOffset) {
+        additionInsertPoint +=
+          changes[changes.length - 1].from +
+          changes
+            .slice(0, changes.length - 1)
+            .flatMap((c) => c.insert)
+            .join("").length +
+          additionPrefix.length;
+      } else {
+        additionInsertPoint =
+          changes[changes.length - 1].from +
+          changes
+            .slice(0, changes.length - 1)
+            .flatMap((c) => c.insert)
+            .join("").length;
+      }
+    } else if (additionPrefix.length) {
+      //if code inserted below existing lines
+      const trailingNewLineChar = 1;
+      additionInsertPoint =
+        changes[changes.length - 1].from +
+        changes[changes.length - 1].insert.length -
+        addition.length -
+        trailingNewLineChar;
+    }
+  }
```

Comment:
@microbit-matt-hillsdon I appreciate that this is absolutely disgusting and that there must be a better way to handle cursor position, but it does seem to work and covers some edge cases that the current dnd doesn't (e.g., drag and drop sound example into an indented block while audio isn't already imported and see where the cursor ends up). I'll leave this here to highlight the various cases we have to cater for. Single and multiline examples still need addressing.

Reply:
Significantly revised in 8ec6f83678acaa9ff62e3750ca28fa6ff952069f with some subsequent tidying.